go/libraries/doltcore/sqle/dsess/gc_safepoint_controller.go

```diff
@@ -0,0 +1,306 @@
+			// session is in a usable state when the callback gets
+			// canceled before ever being called.
+			state.BeginOutstandingVisitCall()
+			go func() {
+				err := visitQuiescedSession(ctx, sess)
+				ret.accumulateErrors(err)
+				ret.wg.Done()
+				state.EndOutstandingVisitCall()
+			}()
+		}
+		if state.OutstandingCommand {
+			// If a command is currently running on the session, register
+			// our work to run as soon as the command is done.
+			state.CommandEndCallback = work
+		} else {
+			// When no command is running on the session, we can immediately
+			// visit it.
+			work()
+		}
+	}
+	return ret
+}
+
+func (w *GCSafepointWaiter) accumulateErrors(err error) {
+	if err != nil {
+		w.mu.Lock()
+		w.err = errors.Join(w.err, err)
+		w.mu.Unlock()
+	}
+}
+
+// |Wait| will block on the Waiter's waitgroup. A successful
+// return from this method signals that all sessions that were known
+// about when the waiter was created have been visited by the
+// |visitQuiescedSession| callback that was given to |Waiter|.
+//
+// This function will return early, and with an error, if the
+// supplied |ctx|'s |Done| channel delivers. In that case,
+// all sessions will not necessarily have been visited, but
+// any visit callbacks which were started will still have
+// completed.
+//
+// In addition to returning an error if the passed in |ctx|
+// is |Done| before the wait is finished, this function also
+// returns accumulated errors as seen from each
+// |visitQuiescedSession| callback. No attempt is made to
+// cancel callbacks or to return early in the case that errors
+// are seen from the callback functions.
+func (w *GCSafepointWaiter) Wait(ctx context.Context) error {
+	done := make(chan struct{})
+	go func() {
+		w.wg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+		return w.err
+	case <-ctx.Done():
+		w.controller.mu.Lock()
+		for _, state := range w.controller.sessions {
+			if state.CommandEndCallback != nil {
+				// Do not visit the session, but do
+				// count down the WaitGroup so that
+				// the goroutine above still completes.
+				w.wg.Done()
+				state.CommandEndCallback = nil
+			}
+		}
+		w.controller.mu.Unlock()
+		// Once a session visit callback has started, we
+		// cannot cancel it. So we wait for all the inflight
+		// callbacks to be completed here, before returning.
+		<-done
+		return errors.Join(context.Cause(ctx), w.err)
+	}
+}
+
+// Beginning a command on a session has three effects:
+//
+// 1) It registers the Session in the set of all
+//    known sessions, |c.sessions|, if this is our
+//    first time seeing it.
+//
+// 2) It blocks for any existing call to |CommandEndCallback|
+//    on this session to complete. If a call to |CommendEndCallback|
+//    is outstanding, our |QuiesceCallbackDone| a read from our
+//    |QuiesceCallbackDone| channel will block.
+//
+// 3) It sets |OutstandingCommand| for the Session to true. Only
+//    one command can be outstanding at a time, and whether a command
```

Comment:
wording maybe confused me a bit, you mean that each session can only have one controller callback at a time?

Reply:
also seems like you use OutstandingVisitCall and OutstandingCommand interchangeably?

Reply:
the optionality and use of this in waiter seems like it's basically like "do callback", which is always going to be a session finalize outside of unit tests. I guess just it might be possible to simplify "SessionCommand"'s naming, there are so many session prefixed names moving around

Reply:
Hmm, I could definitely find better names.

OutstandingVisitCall and OutstandingCommand are different concepts here. Basically, OutstandingCommand just means "the application layer is currently touching this session." OutstandingCommand is used to control when we visit the session if we make a Waiter – we can visit sessions without commands immediately, but for sessions currently doing work, we need to wait until their commandend call comes in.

OutstandingVisitCall means we made a waiter and we are currently calling the callback on this session. In the case, the rest of the application layer should not be touching the session, and SessionCommandBegin will block.